dpctl/tensor/libtensor/include/kernels/elementwise_functions/round.hpp

```diff
@@ -0,0 +1,217 @@
+//
+//                      Data Parallel Control (dpctl)
+//
+// Copyright 2020-2023 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===---------------------------------------------------------------------===//
+///
+/// \file
+/// This file defines kernels for elementwise evaluation of ROUND(x) function.
+//===---------------------------------------------------------------------===//
+
+#pragma once
+#include <CL/sycl.hpp>
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+
+#include "kernels/elementwise_functions/common.hpp"
+
+#include "utils/offset_utils.hpp"
+#include "utils/type_dispatch.hpp"
+#include "utils/type_utils.hpp"
+#include <pybind11/pybind11.h>
+
+namespace dpctl
+{
+namespace tensor
+{
+namespace kernels
+{
+namespace round
+{
+
+namespace py = pybind11;
+namespace td_ns = dpctl::tensor::type_dispatch;
+
+using dpctl::tensor::type_utils::is_complex;
+
+template <typename argT, typename resT> struct RoundFunctor
+{
+
+    // is function constant for given argT
+    using is_constant = typename std::false_type;
+    // constant value, if constant
+    // constexpr resT constant_value = resT{};
+    // is function defined for sycl::vec
+    using supports_vec = typename std::false_type;
+    // do both argTy and resTy support sugroup store/load operation
+    using supports_sg_loadstore = typename std::negation<
+        std::disjunction<is_complex<resT>, is_complex<argT>>>;
+
+    resT operator()(const argT &in)
+    {
+
+        if constexpr (std::is_integral_v<argT>) {
+            return in;
+        }
+        else if constexpr (is_complex<argT>::value) {
+            using realT = typename argT::value_type;
+            return resT{round_func<realT>(std::real(in)),
+                        round_func<realT>(std::imag(in))};
+        }
+        else {
+            return round_func<argT>(in);
+        }
+    }
+
+private:
+    template <typename T> T round_func(T input) const
+    {
+        if (input == 0) {
+            return input;
+        }
+        else if (std::abs(input - std::floor(input)) ==
+                 std::abs(input - std::ceil(input)))
+        {
+            return static_cast<int>(std::ceil(input)) % 2 == 0
+                       ? std::ceil(input)
+                       : std::floor(input);
+        }
+        else {
+            return std::round(input);
+        }
+    }
+};
```

Comment:
Instead of doing this, is there a reason to not use ``std::rint``? It rounds to the nearest integer and rounds ties to even (so 2.5 to 2 and 3.5 to 4).

The only caveat is that the rounding mode is taken into account, but I'm not sure if this would ever be an issue.

There is also a ``sycl::rint`` which does the same, and would enable a ``sycl::vec`` overload.

Reply:
Essentially, the code would simplify to:

```suggestion
    resT operator()(const argT &in)
    {

        if constexpr (std::is_integral_v<argT>) {
            return in;
        }
        else if constexpr (is_complex<argT>::value) {
            return resT{std::rint(std::real(in)),
                        std::rint(std::imag(in))};
        }
        else {
            return std::rint(in);
        }
    }
```

Reply:
I was not aware of `std::rint`. Thanks for introducing it.

Reply:
I would still keep the implementation, even though it is just a call to `std::rint` in the private templated method. These methods are inline and compiler can generate the same code as direct usage, but it keeps the intent clear that `round` acts on complexes elementwise.

Reply:
FYI: https://stackoverflow.com/questions/47156806/stdnearbyint-vs-stdround-in-c11

From https://en.cppreference.com/w/cpp/numeric/math/nearbyint:

> The only difference between std::nearbyint and [std::rint](https://en.cppreference.com/w/cpp/numeric/math/rint) is that std::nearbyint never raises [FE_INEXACT](https://en.cppreference.com/w/cpp/numeric/fenv/FE_exceptions).

Since offloaded kernels never set FP exception flag, in offloaded kernels there is no difference.